File: leetcode-python-go/1604. 警告一小时内使用相同员工卡大于等于三次的人/main.py
from typing import List
from collections import defaultdict

keyName = ["daniel","daniel","daniel","luis","luis","luis","luis"]
keyTime = ["10:00","10:40","11:00","09:00","11:00","13:00","15:00"]

def alertNames(keyName: List[str], keyTime: List[str]) -> List[str]:
    timeMap = defaultdict(list)
    for name, time in zip(keyName, keyTime):
        hour, minute = int(time[:2]), int(time[3:])
        timeMap[name].append(hour * 60 + minute)

    ans = []
    for name, times in timeMap.items():
        times.sort()
        if any(t2 - t1 <= 60 for t1, t2 in zip(times, times[2:])):
            ans.append(name)
    ans.sort()
    return ans

print(alertNames(keyName, keyTime))

File: leetcode-python-go/1604. 警告一小时内使用相同员工卡大于等于三次的人/main.go
package main

import (
	"fmt"
	"sort"
)

func alertNames(keyName, keyTime []string) (ans []string) {
	timeMap := map[string][]int{}
	for i, name := range keyName {
		t := keyTime[i]
		hour := int(t[0]-'0')*10 + int(t[1]-'0')
		minute := int(t[3]-'0')*10 + int(t[4]-'0')
		timeMap[name] = append(timeMap[name], hour*60+minute)
	}
	for name, times := range timeMap {
		sort.Ints(times)
		for i, t := range times[2:] {
			if t-times[i] <= 60 {
				ans = append(ans, name)
				break
			}
		}
	}
	sort.Strings(ans)
	return
}

func main() {
	keyName := []string{"daniel", "daniel", "daniel", "luis", "luis", "luis", "luis"}
	keyTime := []string{"10:00", "10:40", "11:00", "09:00", "11:00", "13:00", "15:00"}
	results := alertNames(keyName, keyTime)
	fmt.Println(results)
}
